### ISSUE TITLE
I use the latest version of the server(0.6.0), not receiving uplink data.

### ISSUE THREAD

**Issue by welt1** [gotthardp/lorawan-server#476]
I've used 0.4.13, 0.5.4, and 0.5.5 versions of server, which are good to use. 

Today, I used the latest version of the server, but found no uplink data, only the downstream packets from server, and the Dashboard window's icon turns blue, as shown in the following figure: 

![dashboard](https://user-images.githubusercontent.com/24259953/45012254-d9064200-b048-11e8-83e6-5f502ab1b57e.PNG)
![frames](https://user-images.githubusercontent.com/24259953/45012269-f1765c80-b048-11e8-9ab8-04d087015be0.PNG)
![ds_frames](https://user-images.githubusercontent.com/24259953/45012273-f6d3a700-b048-11e8-9c55-1985075e2a95.PNG)

From the gateway log, my uplink packets send success,  and received the Ack
from the server.
![gw](https://user-images.githubusercontent.com/24259953/45012376-719cc200-b049-11e8-8f73-72827e019be1.PNG)

My settings for server are consistent with the old version, is there anything that needs to be changed in this version? 

Please help to see, where is the problem? Thank you!

**Comment by gotthardp** [comment]
Did you define some Groups and Areas, reviewed the Gateway and Profile configuration and assigned those to some Group/Area? If yes, please have a look into the error.log of the server.

**Comment by welt1** [comment]
yes, i defined.
I used the original version of 0.60 in July, no problem, the setting is similar to now, also need to define groups and Areas.
error.log only one log:
2018-09-04 11:15:18.211 [error] <0.362.0> node 023CE114 <<"too_early">>

**Comment by gotthardp** [comment]
<<"too_early">> is a downlink related error. Does the server really print nothing upon uplink? The cyan timeline boxes you see are downlinks only (I suppose Class C, right?) There must be a reason for not seeing any uplinks.

**Comment by gotthardp** [comment]
And the LED changes or not? Does the server work and doesn't log, or doesn't even work? But if it sends downlinks then it has to work.

**Comment by welt1** [comment]
Looks all right, just no upstream packet, here is the Debug.log record:

2018-09-04 17:39:15.174 [debug] <0.1432.0>@lorawan_application_semtech_mote:handle_rxq:35 PUSH_DATA <<2,60,225,20>> 4369 0 240
2018-09-04 17:39:27.371 [debug] <0.1435.0>@lorawan_application_semtech_mote:handle_rxq:35 PUSH_DATA <<2,60,225,20>> 4369 0 240
2018-09-04 17:39:36.617 [debug] <0.1438.0>@lorawan_application_semtech_mote:handle_rxq:35 PUSH_DATA <<2,60,225,20>> 4369 0 240
2018-09-04 17:39:46.673 [debug] <0.1441.0>@lorawan_application_semtech_mote:handle_rxq:35 PUSH_DATA <<2,60,225,20>> 4369 0 240
2018-09-04 17:39:58.019 [debug] <0.1446.0>@lorawan_application_semtech_mote:handle_rxq:35 PUSH_DATA <<2,60,225,20>> 4369 0 240
2018-09-04 17:40:03.917 [debug] <0.1449.0>@lorawan_application_semtech_mote:handle_rxq:35 PUSH_DATA <<2,60,225,20>> 4369 0 240

**Comment by gotthardp** [comment]
But that *is* an upstream packet. It seems that only the uplinks are not logged.

**Comment by welt1** [comment]
I also found,  when I send a data format wrong upstream package.
This uplink package will be recorded.
![dashboard1](https://user-images.githubusercontent.com/24259953/45026074-c81cf700-b06f-11e8-9b25-2ab0cbb67ef6.PNG)
![frames1](https://user-images.githubusercontent.com/24259953/45026080-ce12d800-b06f-11e8-9062-1b0e04c7c9fa.PNG)
 Debug.log record:
2018-09-04 18:20:14.642 [error] <0.1792.0> node 023CE114 {not_semtech_mote,{2,<<17,17,0,0,17,17,240,0,0,17,0,0,34,34,34>>}}

**Comment by gotthardp** [comment]
It is really weird. On my installation I can see all uplinks from the mote.

**Comment by gotthardp** [comment]
Could you please "Purge" the frames you have, let the mote sent few frames and then display the frames via `http://localhost:8080/api/rxframes`.

**Comment by welt1** [comment]
Thank you for your help!
I  cleared the original frames, and re-tested it. Let mote send 7 frames, the first 6 correct format data, the last one is the format wrong data. The result is only the last record of the upstream frames.
![dashboard2](https://user-images.githubusercontent.com/24259953/45066287-49b46980-b0f0-11e8-8985-6816d0e2473b.PNG)


![frames2](https://user-images.githubusercontent.com/24259953/45066301-546efe80-b0f0-11e8-983e-fa15f1a0c5d2.PNG)

http://127.0.0.1:8080/api/rxframes?_sortField=datetime&_sortDir=ASC 
![frames_1](https://user-images.githubusercontent.com/24259953/45066336-88e2ba80-b0f0-11e8-9949-9b1d322cd6ed.PNG)

![frames_2](https://user-images.githubusercontent.com/24259953/45066340-8e400500-b0f0-11e8-9473-54afc6b620aa.PNG)

**Comment by welt1** [comment]
![device](https://user-images.githubusercontent.com/24259953/45066407-eb3bbb00-b0f0-11e8-85d4-ea90d25ed714.PNG)

From this view, it is true that the server received 7 upstream data, sending 6 downstream data (the last error format did not). But why is the uplink data not recorded?

**Comment by welt1** [comment]
![mnesia](https://user-images.githubusercontent.com/24259953/45067724-e37f1500-b0f6-11e8-8abd-3fcedf525582.PNG)

This is the rxframe record in Mnesia. From this view, it seems that the server received uplink data at the same time send the downlink data, if there is downlink data to ignore uplink data, only record the downlink data, no downlink data to record the uplink data.

**Comment by gotthardp** [comment]
You are very skilled with mnesia :)
What is the machine / OS you are using? Is it something special? Could you run `erl` and then type `erlang:system_info(os_system_time_source).` and let me know what is the outcome?

I suspect you have insufficient time precision and the frames overwrite each other. Could it be possible on your system?

**Comment by welt1** [comment]
PS F:\work\erlang\temp\lorawan-server> erl
Eshell V9.2  (abort with ^G)
1> erlang:system_info(os_system_time_source).
[{function,'GetSystemTime'},
 {resolution,100},
 {parallel,yes},
 {time,1573058956161024}]
2>

Would that be the problem?

**Comment by gotthardp** [comment]
Yes, I think that's the problem. Your resolution is too small. The "frame id" is generated based on system_time and your is moving too slowly (because of the low resolution) so two frames (uplink and downlink) get the same ID and the later (downlink) overrides the former (uplink).

Just for comparison, I am getting
```
1> erlang:system_info(os_system_time_source).
[{function,clock_gettime},
 {clock_id,'CLOCK_REALTIME'},
 {resolution,1000000000},
 {parallel,yes},
 {time,1536228835314423705}]
```

Your resolution is small. And I also suspect the time will have always 1024 at the end, which means the last byte is always zero.

**Comment by gotthardp** [comment]
Would you do one more test for me, please?

Run the following command:
```erlang
{os:timestamp(), erlang:system_time(millisecond), erlang:system_time(microsecond), erlang:system_time(nanosecond)}.
```

4 times, very quicky after each other (as quickly as you can :) and post here what numbers you got. I consider this as a bug, but need to find a way how to safely generate the ID.

**Comment by welt1** [comment]
Oh, there's a problem, you're right. Here are the test results:

![time](https://user-images.githubusercontent.com/24259953/45191182-6da0b800-b274-11e8-8d25-04ae02a837a2.PNG)

**Comment by gotthardp** [comment]
Thanks for the tests. I made an update, which should fix this problem. This will be released as 0.6.0 in the coming days.

**Comment by welt1** [comment]
Testing the new version is OK, this problem has been solved, thank you very much!
